xls/modules/rle/rle_dec.x

```diff
@@ -0,0 +1,520 @@
+// next evaluation      XXXXXX║ 0   ║ 1   ║ 2   ║ 3   ║ 4   ║ 5   ║ 6   ║ ...
+//                      ──────╨─────╨─────╨─────╨─────╨─────╨─────╨─────╨────
+// do_recv                    ┌─────┐     ┌─────────────────┐           ┌────
+//                      ──────┘     └─────┘                 └───────────┘
+//                      ──────╥─────╥─────╥─────╥─────╥─────╥────────────────
+// symbol, count        XXXXXX║ A,2 ║XXXXX║ B,1 ║ B,1 ║ C,3 ║XXXXXXXXXXXXXXXX
+// (input channel)      ──────╨─────╨─────╨─────╨─────╨─────╨────────────────
+// last                                   ┌─────┐     ┌─────┐
+// (input channel)      ──────────────────┘     └─────┘     └────────────────
+//                      ╥─────╥───────────╥───────────╥─────────────────╥────
+// state.symbol         ║ 0   ║ A         ║ B         ║ C               ║ 0
+// (set state value)    ╨─────╨───────────╨───────────╨─────────────────╨────
+//                      ╥─────╥─────╥─────╥───────────╥─────╥─────╥─────╥────
+// state.count          ║ 0   ║ 1   ║ 0   ║ 0         ║ 2   ║ 1   ║ 0   ║ 0
+// (set state value)    ╨─────╨─────╨─────╨───────────╨─────╨─────╨─────╨────
+//
+//                      ──────╥───────────╥───────────╥─────────────────╥────
+// symbol               XXXXXX║ A         ║ B         ║ C               ║XXXX
+// (output channel)     ──────╨───────────╨───────────╨─────────────────╨────
+// last                                   ┌─────┐                 ┌─────┐
+// (output channel)     ──────────────────┘     └─────────────────┘     └────
+
+
+import std
+import xls.modules.rle.rle_common as rle_common
+
+type DecInData  = rle_common::EncData;
+type DecOutData = rle_common::PlainData;
+
+// structure to preserve the state of an RLE decoder
+struct RunLengthDecoderState<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+  // symbol to be repeated on output
+  symbol: bits[SYMBOL_WIDTH],
+  // count of symbols that has to be send
+  count: bits[COUNT_WIDTH],
+  // send last when repeat ends
+  last: bool,
+}
+ // RLE decoder implementation
+pub proc RunLengthDecoder<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+  input_r: chan<DecInData<SYMBOL_WIDTH, COUNT_WIDTH>> in;
+  output_s: chan<DecOutData<SYMBOL_WIDTH>> out;
+
+  init {(
+    RunLengthDecoderState<SYMBOL_WIDTH, COUNT_WIDTH> {
+      symbol: bits[SYMBOL_WIDTH]:0,
+      count:  bits[COUNT_WIDTH]:0,
+      last:   bool:false,
+    }
+  )}
+
+  config (
+    input_r: chan<DecInData<SYMBOL_WIDTH, COUNT_WIDTH>> in,
+    output_s: chan<DecOutData<SYMBOL_WIDTH>> out,
+  ) {(input_r, output_s)}
+
+  next (tok: token, state: RunLengthDecoderState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+    let zero_input = DecInData {
+      symbol: bits[SYMBOL_WIDTH]:0,
+      count: bits[COUNT_WIDTH]:0,
+      last: false
+    };
+    let recv_next_symbol = (state.count == bits[COUNT_WIDTH]:0);
+    let (tok, input) = recv_if(tok, input_r, recv_next_symbol, zero_input);
+    let (next_symbol, count, next_last) = if (recv_next_symbol) {
```

Comment:
I wonder if we could simply assign the `zero_input` so the current state and get rid of this if?

Reply:
Changed it